crates/compiler/parse/src/expr.rs

```diff
@@ -2696,48 +2696,154 @@ impl<'a> Spaceable<'a> for RecordField<'a> {
+    use crate::both;
     use RecordField::*;
 
     map_with_arena!(
         and!(
             specialize_err(|_, pos| ERecord::Field(pos), loc!(lowercase_ident())),
             and!(
-                spaces(),
-                optional(either!(
-                    and!(byte(b':', ERecord::Colon), record_field_expr()),
-                    and!(
-                        byte(b'?', ERecord::QuestionMark),
-                        spaces_before(specialize_err_ref(ERecord::Expr, loc_expr(false)))
-                    )
+                optional(both!(
+                    both!(spaces(), byte(b':', ERecord::Colon)),
+                    record_field_expr()
+                )),
+                optional(both!(
+                    both!(spaces(), byte(b'?', ERecord::QuestionMark)),
+                    spaces_before(specialize_err_ref(ERecord::Expr, loc_expr(false)))
                 ))
             )
         ),
-        |arena: &'a bumpalo::Bump, (loc_label, (spaces, opt_loc_val))| {
+        |arena: &'a bumpalo::Bump,
+         (loc_label, opt_loc_val): (
+            Loc<&'a str>,
+            (
+                Option<((&'a [CommentOrNewline<'a>], ()), RecordFieldExpr<'a>)>,
+                Option<((&'a [CommentOrNewline<'a>], ()), Loc<Expr<'a>>)>
+            )
+        )| {
             match opt_loc_val {
-                Some(Either::First((_, RecordFieldExpr::Value(loc_val)))) => {
-                    RequiredValue(loc_label, spaces, arena.alloc(loc_val))
+                (Some((spaces, RecordFieldExpr::Value(loc_val))), None) => {
+                    RequiredValue(loc_label, spaces.0, arena.alloc(loc_val))
                 }
 
-                Some(Either::First((_, RecordFieldExpr::Apply(arrow_spaces, loc_val)))) => {
-                    ApplyValue(loc_label, spaces, arrow_spaces, arena.alloc(loc_val))
+                (Some((spaces, RecordFieldExpr::Apply(arrow_spaces, loc_val))), None) => {
+                    ApplyValue(loc_label, spaces.0, arrow_spaces, arena.alloc(loc_val))
                 }
 
-                Some(Either::Second((_, loc_val))) => {
-                    OptionalValue(loc_label, spaces, arena.alloc(loc_val))
+                (None, Some((spaces, loc_val))) => {
+                    OptionalValue(loc_label, spaces.0, arena.alloc(loc_val))
                 }
 
                 // If no value was provided, record it as a Var.
                 // Canonicalize will know what to do with a Var later.
-                None => {
-                    if !spaces.is_empty() {
-                        SpaceAfter(arena.alloc(LabelOnly(loc_label)), spaces)
-                    } else {
-                        LabelOnly(loc_label)
-                    }
-                }
+                (None, None) => LabelOnly(loc_label),
+
+                (Some(_), Some(_)) => todo!(),
             }
         }
     )
+    .trace("record_field")
+}
+
+#[macro_export]
+macro_rules! both {
```

Comment:
There's already an `and!` macro - how is `both!` different?

Reply:
I agree, this is not ideal, or idiomatic to the rest of parsing. I would like to find a better solution. The difference between `and` and `both` is that `both` is more conservative on returning `Ok`. I'm still a bit fuzzy on the reasoning between `Progress` and `Ok`/`Err`. 

I was getting error because the first `optional` would make some progress, but still return an `Err` and eat part of the state that the second `optional` parser would need. I would also get a similar issue with `one_of`.

Do you have a suggestion on a different way to go about this? I'm not sure how much experience you've got with the parser. I'd love a better way to do this!

Reply:
I tried out `backtrackable` and that seems like it does what I wanted, but I still feel like this parser doesn't quite fit in with the style of the rest of them.